cafebabel/static/js/edit-article.js

```diff
@@ -0,0 +1,147 @@
+class Tags {
+  constructor() {
+    const domContext = document.querySelector('.tags')
+    this.list = domContext.querySelector('.tags-list')
+    this.values = Array.from(this.list.querySelectorAll('input')).map(
+      input => input.value
+    )
+  }
+  addRemoveListener() {
+    const tagsButtonRemove = document.querySelectorAll('.tags .tags-list li a')
+    tagsButtonRemove.forEach(tagButtonRemove =>
+      Listener.clickRemove(tagButtonRemove)
+    )
+  }
+  checkSubmission(query) {
+    return query && !this._isTag(query)
+  }
+  add(query) {
+    this.values = this.values.concat(query)
+  }
+  remove(query) {
+    this.values = this.values.filter(value => value !== query)
+  }
+  display() {
+    const container = this.list.cloneNode(false)
+    this.list.replaceWith(this._createTagsList(container, this.values))
+    this.addRemoveListener()
+  }
+  _isTag(tag) {
+    return this.values.includes(tag)
+  }
+  _createTagsList(container, tagValues) {
+    tagValues.forEach((tagValue, index) =>
+      container.appendChild(this._createTag(tagValue, index))
+    )
+
+    return container
+  }
+  _createTag(tagValue, index) {
+    const li = document.createElement('li')
+    const input = document.createElement('input')
+    const a = document.createElement('a')
+    input.type = 'hidden'
+    input.name = `tag-${index + 1}`
+    input.setAttribute('list', 'tags')
+    input.value = tagValue
+    li.append(tagValue, input, a)
+
+    return li
+  }
+}
+
+class Effect {
+  static fadeIn(element) {
+    function fade() {
+      element.style.opacity = +element.style.opacity + 0.03
+      if (element.style.opacity <= 1) {
+        requestAnimationFrame(fade)
+      }
+    }
+
+    element.style.opacity = 0
+    fade()
+  }
+  static fadeOut(element) {
+    function fade() {
+      element.style.opacity = +element.style.opacity - 0.03
+      if (element.style.opacity < 0) {
+        element.style.display = 'none'
+        resolve()
+      } else {
+        requestAnimationFrame(fade)
+      }
+    }
+
+    element.style.opacity = 1
+
+    return new Promise((resolve, reject) => {
+      fade()
+    })
+  }
+}
+
+class Listener {
+  static clickAdd() {
+    const tagButtonAdd = document.querySelector('.tags button.add')
+    tagButtonAdd.addEventListener('click', event => {
+      event.preventDefault()
+      const submission = event.target.previousSibling.value
+      const tags = new Tags()
+      if (tags.checkSubmission(submission)) tags.add(submission)
+      tags.display()
+      Effect.fadeIn(document.querySelector('.tags .tags-list li:last-child'))
+      document.querySelector('.tags input[name=tag-new]').value = ''
+    })
+  }
+  static clickRemove(tagButtonRemove) {
+    tagButtonRemove.addEventListener('click', event => {
+      const tags = new Tags()
+      const tagElement = event.target.parentNode
+      tags.remove(tagElement.innerText)
+      Effect.fadeOut(tagElement).then(response => tags.display())
```

Comment:
I think you don't have a `response` here given that you don't `resolve()` any value, I would just do `.then(tags.display)`